### PR TITLE
Only submit survey responses for individuals, not programming pairs

### PR DESCRIPTION
**Pull request by jmkulwik**
In the past, when users who were pair programming would take an quiz or survey, their response would be submitted for both partners. Instead, users should only be able to submit responses for themselves. This updates the user model so pair programming is avoided on survey and quiz levels.

## Links

- [jira](https://codedotorg.atlassian.net/browse/LP-207)

## Testing story

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -1772,7 +1773,20 @@ def self.track_level_progress(user_id:, level_id:, script_id:, new_result:, subm
       user_level.atomic_save!
     end
 
-    if pairing_user_ids
+    should_track_pairing = true
+    current_level = user_level.level
+    if current_level.type == "LevelGroup" && current_level.properties["anonymous"] == "true"
```
Comment (jmkulwik) [inline dashboard/app/models/user.rb]: Move to the level model

File: dashboard/app/models/user.rb [hunk 2]
```diff
@@ -1772,7 +1773,20 @@ def self.track_level_progress(user_id:, level_id:, script_id:, new_result:, subm
...
+      should_track_pairing = false
+    end
+
+    if should_track_pairing
+      group_levels = script.levels.select {|level| level.type == "LevelGroup"}
```
Comment (jmkulwik) [inline dashboard/app/models/user.rb]: See if there is another way to access a parent of a level.

File: dashboard/app/models/script.rb [hunk 3]
```diff
@@ -1726,6 +1726,20 @@ def get_assessment_script_levels
     end
   end
 
+  # Determine whether a level in a script should allow pairing based on its
+  # parent level.
+  # TODO: "LevelGroup" levels currently don't have access to their parent level
+  # once they do, simply access the parent level as part of the should_allow_pairing?
+  # check in level.rb
```
Comment (jmkulwik) [inline dashboard/app/models/script.rb]: This functionality will be added by this PR: https://github.com/code-dot-org/code-dot-org/pull/35827

File: dashboard/app/models/levels/level.rb [hunk 4]
```diff
@@ -275,6 +275,10 @@ def write_custom_level_file
     end
   end
 
+  def should_allow_pairing?
+    type != "LevelGroup"
```
Comment (clareconstantine) [inline dashboard/app/models/levels/level.rb]: Should this be in a constant somewhere? it looks like we refer to it as just the string in other places, but might be worth looking for a good place for it to live.

File: dashboard/test/models/level_test.rb [hunk 5]
```diff
@@ -120,6 +120,11 @@ def raise_unless_specifies_ideal_level_source(level_class)
     assert_not custom_levels.include?(@level)
   end
 
+  test "should not allow pairing with levelgroup type levels" do
```
Comment (clareconstantine) [inline dashboard/test/models/level_test.rb:123]: do we need a test for children of LevelGroups?

**Review by mvkski — APPROVED** [review]
LGTM

**Review by clareconstantine — APPROVED** [review]

**Review by clareconstantine — APPROVED** [review]
Looks good!

**Review by cforkish — APPROVED** [review]
LGTM!

File: dashboard/app/models/levels/level.rb (hunk 4 above)
Comment (jmkulwik) [reply dashboard/app/models/levels/level.rb]: This is a good call out. I'm going to leave this as is for now as this PR has been delayed for a while for various reasons, but I'll keep an eye out for a good opportunity to make this change soon.

File: dashboard/app/models/script.rb (hunk 3 above)
Comment (jmkulwik) [reply dashboard/app/models/script.rb]: ✔️

**Pull request merged**